### DISCUSSION TITLE
Securing Anon Access with a Custom API Key in Next.js

### DISCUSSION THREAD

**Discussion by Felix-Schoenherr** [supabase/supabase#30192] [Questions] [answered]
Hey everyone,

I'm currently working on a project where my application relies solely on the anon rule, as I'm not creating any authenticated users. In my Row Level Security (RLS) settings, I’ve granted SELECT access to the anon user. However, I'm concerned about data security, since anyone could potentially fetch data using the public anon key.

### Project Setup
I'm using Next.js and loading the data on the server side. To make the app more secure, I attempted to follow [this Supabase tutorial](https://supabase.com/docs/guides/api/securing-your-api?queryGroups=pre-request&pre-request=use-additional-api-key#examples) on securing your API. The goal is to add a custom API key to control whether requests are allowed, and since I'm server-side rendering, I should be able to keep this key secure.

### The Issue
Unfortunately, I haven't been able to get the tutorial to work as expected. The custom API key doesn't seem to restrict access properly, and I’m not sure what I'm missing.

Could anyone provide guidance or tips on how to successfully implement this? Any help would be greatly appreciated!

**Comment by GaryAustin1** [comment]
You probably need to show what you did with that guide and describe what is not working.
Check the Postgres logs for any error.
It should error if your extra header is not set to key in the extra header.

You can also use `raise log 'api key header = %',req_app_api_key;` to see if the function is being called and got the correct info.

**Reply by Felix-Schoenherr** [reply]
Thank you for helping me. I used you log to determine if the function gets called. I noticed that it did: Event message
api key header = NULL or if i send the correct api key the key is read correctly. 

Then i tried to log which role the user has as i noticed that the function is stopped pretty fast. (Sorry for not doing this earlier, but I am pretty new to POSTgres and Supabase).

`raise log 'role = %', current_role;`
`role = postgres`

as you can see the request hast the role postgres. But I am not sure why. 

My server.ts as recommended in the supabase guide:

`import { createServerClient, type CookieOptions } from '@supabase/ssr';
import { cookies } from 'next/headers';

export async function createClient() {
    const cookieStore = await cookies();

    return createServerClient(process.env.NEXT_PUBLIC_SUPABASE_URL!, process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!, {
        cookies: {
            getAll() {
                return cookieStore.getAll();
            },
            setAll(cookiesToSet) {
                try {
                    cookiesToSet.forEach(({ name, value, options }) => cookieStore.set(name, value, options));
                } catch {
                    // The `setAll` method was called from a Server Component.
                    // This can be ignored if you have middleware refreshing
                    // user sessions.
                }
            },
        },
        global: {
            headers: {
                'x-app-api-key': process.env.SUPABASE_SERVER_KEY!,
            },
        },
    });
}
`

and this is one of my requests:

`'use server';

import { createClient } from '@/utils/supabase/server';

export const fetchAllEvents = async () => {
    const supabase = await createClient();
    const { data, error } = await supabase.from('event').select('*');

    if (error) {
        console.error('Error fetching all events:', error);
        return [];
    }

    return data || [];
};
`

**Reply by GaryAustin1** [reply, marked as answer]
The function would have postgres role as it is defined as `security definer` type.
I think that must be a bug in Supabase's example.

This example from PostgREST (The REST API used) does not have it as security definer.
https://docs.postgrest.org/en/stable/references/transactions.html#setting-headers-via-pre-request

You also might be able to get the calling role with:
`calling_role text := current_setting('request.headers', true)::json->>'role';`

**Reply by Felix-Schoenherr** [reply]
Yes you are right, thanks for helping!  

Altough this line is not working:` calling_role text := current_setting('request.headers', true)::json->>'role';`
The correct syntax is: `calling_role := current_setting('request.jwt.claims', true)::json->>'role';`

If somebody is curious on how i fixed it:


```plpgsql
declare
  req_app_api_key text := current_setting('request.headers', true)::json->>'x-app-api-key';
  is_app_api_key_registered boolean;
  calling_role text;
begin
  calling_role := current_setting('request.jwt.claims', true)::json->>'role';
  if calling_role <> 'anon' then
    -- not `anon` role, allow the request to pass
    return;
  end if;
  select
    true into is_app_api_key_registered
  from private.anon_api_keys
  where
    id = req_app_api_key
  limit 1;

  if is_app_api_key_registered is true then
    -- api key is registered, allow the request to pass
    return;
  end if;
  raise sqlstate 'PGRST' using
    message = json_build_object(
      'message', 'No registered API key found in x-app-api-key header.')::text,
    detail = json_build_object(
      'status', 403)::text;
end;
```

**Reply by GaryAustin1** [reply]
Sorry about the headers versus jwt.claims... was a quick copy/edit.  Glad you understood.
I submitted a doc PR to Supabase to change the role variable and cast the key to UUID as their API key table is UUID.

**Comment by Flowsmooth** [comment]
Thank you.


Sent from Yahoo Mail for iPhone


On Thursday, October 31, 2024, 3:59 AM, Felix ***@***.***> wrote:




Yes you are right, thanks for helping!

Altough this line is not working: calling_role text := current_setting('request.headers', true)::json->>'role';
The correct syntax is: calling_role := current_setting('request.jwt.claims', true)::json->>'role';

If somebody is curious on how i fixed it:

declare req_app_api_key text := current_setting('request.headers', true)::json->>'x-app-api-key'; is_app_api_key_registered boolean; calling_role text; begin calling_role := current_setting('request.jwt.claims', true)::json->>'role'; if calling_role <> 'anon' then -- not anon` role, allow the request to pass
return;
end if;
select
true into is_app_api_key_registered
from private.anon_api_keys
where
id = req_app_api_key
limit 1;

if is_app_api_key_registered is true then
-- api key is registered, allow the request to pass
return;
end if;
raise sqlstate 'PGRST' using
message = json_build_object(
'message', 'No registered API key found in x-app-api-key header.')::text,
detail = json_build_object(
'status', 403)::text;
end;`
